### ISSUE TITLE
Change visibility of function "tryToGetAvailableIdProductAttribute" in ProductController

### ISSUE THREAD

**Issue by cdcvince** [PrestaShop/PrestaShop#18901]
Hello,

I would like to override the behavior of the private function in **ProductController.php** :
```
private function tryToGetAvailableIdProductAttribute($checkedIdProductAttribute)
```
Can you change the visibility to 'protected' ?

This function is used for :

```
    /**
     * If the PS_DISP_UNAVAILABLE_ATTR functionality is enabled, this method check
     * if $checkedIdProductAttribute is available.
     * If not try to return the first available attribute, if none are available
     * simply returns the input.
     *
     * @param int $checkedIdProductAttribute
     *
     * @return int
     */
```

The use case of this override is pretty straightforward, it returns the next available combination when the selected combination is not available. But maybe we don't want the default combination returned by Prestashop, but a smart mechanism to return the best combination (based on external data for example).

**Comment by florine2623** [comment]
Hi @cdcvince ,

You can override this file by recreating it if the` /override` folder and putting your method in the **ProductController.php** file.

If you need customized development, you can contact a PrestaShop Agency : 
https://www.prestashop.com/en/experts?iso=&budgetMin=0&budgetMax=50000

Thanks!

**Comment by cdcvince** [comment]
Thank you @florine2623 for your answer but we cannot override a **private** function. That's exactly for what private is made for.
We can only override function with **public** or **protected** visibility. Which is exactly what I asked.

**Comment by eternoendless** [comment]
You're right @cdcvince. There's a problem because you can't customize that part.

This is symptomatic of a design problem in that controller, because it does too much. However, I'm not a fan of making private methods protected, firstly because it violates the open/closed principle of SOLID design, and secondly (and most importantly) because it would promote that method as part of the API, rendering it impossible to refactor without introducing a breaking change.

My first reaction would be to suggest adding a hook there. But I don't think it's the right choice. Hooks are generally meant to allow enriched behavior, not _replace_ core behavior.

The best solution in my opinion would be to refactor these methods by extracting that logic into a service. That way, the service could be decorated/replaced as needed by modules.
